EIPS/eip-fhe.md

```diff
@@ -0,0 +1,55 @@
+Fully homomorphic encryption allows anyone to bootstrap a "storage-key" for their data which allows any endparty to execute mutually agreed boolean circuits on
+encrypted data, without decrypting it, including searching and indexing too. This allows secure storage and operations on healthcare records or customer databases, enabling services easy compliance with EU GDPR. Concensus about data modification will still be transparent for storage keys are stored in the contracts.
+
+## Motivation
+A very common problem among people who wants to migrate or enter their business to blockchains is that they cannot put confidential data into a public ledger,
+although they still need to overcome distrust between different types of customers. Uploading critical business data into a blockchain where competitors or unauthorized personnel can scrape and analyze it is not an option so in all cases these parties stay with the centralized non-transparent model.
+This EIP issues this problem with enabling distribution of existing infrastructures on the blockchain, without exposing sensitive data into publicity.
+
+## Specification
+
+Contract storages will have a subset (or extension?) dedicated for encrypted data with a map of storage-keys on which can operate on a specified offset range.
+Contracts can read each other's storage using a known storage key so multiparty confidential storage is possible with for eg `web3.eth.get(Secret?)StorageAt(...)`.
+Bootstrapping is implemented using learning with errors which is known to be as hard as lattice shortest-vector problems (are quantum-resistant by design),
+and allow near-infinite amounts of operations on secrets.
+
+Implementation should include a flag of LWE-depth and noise `n`. Boolean primitives `x` are handled as gadget matrices (generated with a decomposition
+function `D(v))` with the restriction that there exists an LWE sample outputing a small vector `u` where `ux -v` are less or equal to an agreed precision (see Gentry-Saha-Waters).
+Decomposition algorithms shall include entropy to keep error limits independent.
+Operating on encrypted data with these gadgets are possible with the use of a key-switching context which will output another sample with a different key
+on the same data with only negligible difference on the noise parameter, which enables us billions of consequential switches.
+Specifically the key-switching takes k LWE sample ciphertexts, an R-lipschitzian morphism of Z-modules, a key sample KS with noise limit `l`, and `Aks` and `Vks`
+being the amplitude and variance of error, the algorithm shall output:
+`||error(c)||∞ <= R||error(c)||∞  + nt*n*Aks + n*2^-(t+1)` in worst case and
+`variance(error(c)) <= R^2*variance(error(c)) + nt*n*Vks + n*2^(2*(t+1))` in average case
+
+
+Most of the design is based on the academic works of Ilaria Chillotti1, Nicolas Gama, Mariya Georgieva, and Malika Izabachene, cryptology ePrint 2017/430
```

Comment:
Gas costs need to be specified eventually.  Are there some measurements of memory/cpu consumption?

Reply:
Core i7-4910MQ laptop, x86_64, assuming parameters to achieve 110bit security:
* Circuit bootstrap 137ms 
* Fresh CMux tXP = 34µs
* CB CMux 34µs
* PubKST BSR tKS = 180ms
worst-case noise addition with these parameters: 2^-20
No measurments on memory usage yet